File: Proxy/StaticProxy.java
interface Subject {
    void doSomething();
}

class ConcreteSubject implements Subject {
    public void doSomething() {
        System.out.println("do something");
    }
}

class Proxy implements Subject {
    private Subject target;

    public Proxy(Subject target) {
        this.target = target;
    }

    public void doSomething() {
        System.out.println("before");
        target.doSomething();
        System.out.println("after");
    }
}

class StaticProxy {
    public static void main(String[] args) {
        Proxy proxy = new Proxy(new ConcreteSubject());
        proxy.doSomething();
    }
}

File: Proxy/StaticProxy.py
from abc import ABCMeta, abstractmethod

class Subject:
    __metaclass__ = ABCMeta

    @abstractmethod
    def do_something(self):
        pass

class ConcreteSubject(Subject):
    def do_something(self):
        print("do something")

class Proxy(Subject):
    def __init__(self, target):
        self._target = target

    def do_something(self):
        print("before")
        self._target.do_something()
        print("after")

if __name__ == "__main__":
    proxy = Proxy(ConcreteSubject())
    proxy.do_something()

